db/filters/base.py

```diff
@@ -0,0 +1,381 @@
+"""
+This module contains Predicate subclasses, and relevenat mixins, that describe nodes in a
+predicate tree, or, in other words, predicates that compose into a tree. A predicate is
+described by whether it's a Leaf or a Branch, whether it takes parameters and how many
+(SingleParameter, MultiParameter, NoParameter) and whether it relies on comparability
+(ReliesOnComparability), as well as its identifier and its human-friendly name.
+"""
+
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import Any, List, Union, Type, Set
+from sqlalchemy_filters.exceptions import BadFilterFormat as SABadFilterFormat
+from sqlalchemy import column, not_, and_, or_
+from abc import ABC, abstractmethod
+
+
+class PredicateSuperType(Enum):
+    """Every Predicate is either a leaf node in the predicate tree, or a branch node.
+    A leaf node (e.g. EMPTY, EQUAL) has no predicate children, while a branch node
+    (e.g. AND, OR, NOT) always has predicate children."""
+    LEAF = "leaf"
+    BRANCH = "branch"
+
+
+class BranchPredicateType(Enum):
+    NOT = "not"
+    OR = "or"
+    AND = "and"
+
+
+class LeafPredicateType(Enum):
```

Comment:
It's for enumerating all leaf predicate types. It's not strictly needed, since there's a list of available predicate classes which we can query. An enum is convenient when you want to check if a predicate type string is valid, or whether it's a leaf or a branch.